packages/drawer/src/ux-modal-service.ts

```diff
@@ -0,0 +1,277 @@
+    const bindingContext: ModalBindingContext = {};
+    if (!options.viewModel && !options.view) {
+      throw new Error('Invalid Drawer Settings. You must provide "viewModel", "view" or both.');
+    }
+    const element = document.createElement('ux-drawer');
+    element.setAttribute('dismiss.delegate', 'dismiss()');
+    element.setAttribute('ok.delegate', 'ok($event)');
+    if (options.position !== undefined) {
+      element.setAttribute('position', options.position);
+    }
+    if (options.moveToBodyTag === false) {
+      element.setAttribute('move-to-body-tag.bind', 'false');
+    }
+    if (options.overlayDismiss === false) {
+      element.setAttribute('overlay-dismiss', 'false');
+    }
+    if (options.keyboard !== undefined) {
+      bindingContext.keyboard = options.keyboard;
+      element.setAttribute('keyboard.bind', 'keyboard');
+    }
+    if (typeof options.restoreFocus === 'function') {
+      bindingContext.restoreFocus = options.restoreFocus;
+      element.setAttribute('restore-focus.bind', 'restoreFocus');
+    }
+    
+    element.setAttribute('host.bind', 'false');
+    
+    if (options.theme) {
+      bindingContext.theme = options.theme;
+      element.setAttribute('theme.bind', `theme`);
+    }
+    const compose = document.createElement('compose');
+    compose.setAttribute('view-model.ref', 'composeViewModel');
+    element.innerHTML = compose.outerHTML;
+    if (!options.host || options.host === 'body') {
+      options.host = document.body;
+    } else if (typeof options.host === 'string') {
+      options.host = document.querySelector(options.host) || document.body;
+    }
+    options.host.appendChild(element);
+    let childView = this.templatingEngine.enhance({ element: element, bindingContext: bindingContext });
+    
+    if (options.viewModel && bindingContext.composeViewModel) {
+      bindingContext.composeViewModel.viewModel = options.viewModel;
+    }
+    if (options.view && bindingContext.composeViewModel) {
+      bindingContext.composeViewModel.view = options.view;
+    }
+    if (options.model && bindingContext.composeViewModel) {
+      bindingContext.composeViewModel.model = options.model;
+    }
+
+    const controllers = (childView as any).controllers as Controller[];
+    
+    const drawer: UxDrawer =  controllers[0].viewModel as UxDrawer;
+    const drawerIndex = this.drawerIndex;
+    const whenClosed: Promise<ModalServiceResult> = new Promise((resolve) => {
+      this.eventAggregator.subscribeOnce(`drawer-${drawerIndex}-resolve`, (result: ModalServiceResult) => {
+        resolve(result);
+      });      
+    });
+
+    (drawer as UxDrawer & ModalServiceDrawer).whenClosed = () => {
+      return whenClosed;
+    }
+    bindingContext.dismiss = () => {
+      drawer.detached();
+      const parent = drawer.element.parentNode;
+      if (!parent) { return; }
+      parent.removeChild(drawer.element);
+      this.eventAggregator.publish(`drawer-${drawerIndex}-resolve`, {
```

Comment:
this event aggregator will be potentially shared by many drawers, there need to be more information to identify which drawer published this event, or we just dispatch event on the host of this drawer

Reply:
I did it with the `drawerIndex` thinking that there only should be one `ModalService` per app (being a singleton). However, I can also inject a transient instance of `EventAggregagor` in this service which then would not be shared with any other parts of the application. Would be safer.